File: frontend/src/api/accounts.ts
import { api } from "@/api/client";
import type { AccountInput, AccountMoveDirection, AccountWithBalance } from "@/types";

export function fetchAccounts(includeArchived = false) {
  return api.get<AccountWithBalance[]>(`/accounts?include_archived=${includeArchived}`);
}

export function createAccount(input: AccountInput) {
  return api.post<AccountWithBalance>("/accounts", input);
}

export function updateAccount(id: number, input: Partial<AccountInput> & { is_archived?: boolean }) {
  return api.patch<AccountWithBalance>(`/accounts/${id}`, input);
}

export function moveAccount(id: number, destination: AccountMoveDirection | number, includeArchived: boolean) {
  const target = typeof destination === "number" ? { target_account_id: destination } : { direction: destination };
  return api.post<void>(`/accounts/${id}/move`, { ...target, include_archived: includeArchived });
}

export function deleteAccount(id: number) {
  return api.delete<void>(`/accounts/${id}`);
}

File: backend/app/api/routes/accounts.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_session
from app.schemas.account import AccountCreate, AccountMove, AccountUpdate, AccountWithBalance
from app.services.account_service import create_account, delete_account, list_accounts, move_account, update_account

router = APIRouter(prefix="/accounts", tags=["accounts"])


@router.get("", response_model=list[AccountWithBalance])
async def read_accounts(
    include_archived: bool = False, session: AsyncSession = Depends(get_session)
) -> list[AccountWithBalance]:
    return await list_accounts(session, include_archived)


@router.post("", response_model=AccountWithBalance, status_code=201)
async def create_account_route(
    payload: AccountCreate, session: AsyncSession = Depends(get_session)
) -> AccountWithBalance:
    return await create_account(session, payload)


@router.patch("/{account_id}", response_model=AccountWithBalance)
async def update_account_route(
    account_id: int, payload: AccountUpdate, session: AsyncSession = Depends(get_session)
) -> AccountWithBalance:
    return await update_account(session, account_id, payload)


@router.post("/{account_id}/move", status_code=204)
async def move_account_route(
    account_id: int, payload: AccountMove, session: AsyncSession = Depends(get_session)
) -> None:
    await move_account(session, account_id, payload)


@router.delete("/{account_id}", status_code=204)
async def delete_account_route(account_id: int, session: AsyncSession = Depends(get_session)) -> None:
    await delete_account(session, account_id)
